### ISSUE TITLE
how to train it on my own dataset

### ISSUE THREAD

**Issue by derek-zr** [zhaoweicai/cascade-rcnn#3]
hi! I want to train cascade-rcnn on my own dataset (three classes). I don't know how to modify the files(eg. examples/voc/). Can you give me some instructions? Thank you!

**Comment by makefile** [comment]
Hi, When I train the models such as `res50-12s-600-rfcn-cascade` without FPN  with my own dataset is fine. But when I try to train `res50-15s-800-fpn-cascade` with my own dataset, I meet the problem that decode_bbox_layer cannot get valid bbox.  After the code of "screen out high IoU boxes, to remove redundant gt boxes" the valid_bbox_ids is 0.
So, what the problem might be? Thanks. @zhaoweicai

**Comment by zhaoweicai** [comment]
@makefile If you don't want to remove the redundant gt boxes, you can simply set gt_iou_thr=1.0 or higher. But a more important problem is you might not have enough proposals. In your case of error, only gt boxes and no negative box. You can try to lower the proposal threshold in "BoxGroupOutput" layer to have more proposals. Or your training is diverging and crashed. You can also try to use a lower learning rate.

**Comment by makefile** [comment]
@zhaoweicai Thanks! Follow your advice, set lower the fg_thr in BoxGroupOutput layer, the problem disappeared.

**Comment by jwnsu** [comment]
The error seems to relate to multiple gpus. When I tried single gpu, training proceeds; however, with 2 gpus, encountered same above error.

**Comment by makefile** [comment]
@Peng-wei-Yu try lower the score of `fg_thr` instead of nms thresh.

**Comment by Peng-wei-Yu** [comment]
@jwnsu @makefile Thank you for you help. But I tried to lower fg_thr and use only GPU 1, the problem is still there. Have you tried to change the --weights in train_detection, I decided to change the caffemodel and have a try.

**Comment by jwnsu** [comment]
FYI. coco model seems to work fine (e.g. coco/res50-15s-800-fpn-cascade is fine, res101 runs out of GPU memory on 1080 Ti), suggest you switch to coco flavor.

**Comment by zhaoweicai** [comment]
@Peng-wei-Yu when you change the number of GPUs, you should change the learning rate at the same time, as described in the [paper](https://arxiv.org/pdf/1706.02677.pdf).

**Comment by zhaoweicai** [comment]
@jwnsu The code should have no problem on multi-gpu training or VOC dataset. Try the run the script a couple of times to see if the problem still happens. If the problem is still there, try to lower the learning rate a little bit. If it still cannot be fixed, maybe there is something wrong.

**Comment by Peng-wei-Yu** [comment]
@makefile @zhaoweicai When you trained cascade rcnn on your own data, which caffemodel did you use. Your own caffemodel or ResNet-50-model-merge.caffemodel. The picture in my own data have the size of 1600*1200, should I change the short_size and long_size in train.prototxt.

**Comment by makefile** [comment]
@Peng-wei-Yu If you use the author's prototxt, you should use the corresponding ResNet-50-model-merge.caffemodel, since it merges the scale layer to BN layer and freeze the BN layer to reduce memory and speed up. You can increase the input size of image if your memory is enough, but the result may not increase too much.

**Comment by Peng-wei-Yu** [comment]
@makefile Thank you very much. I'll have a try by using ResNet-50-model-merge.caffemodel.

**Comment by GuoxingYan** [comment]
@makefile @Peng-wei-Yu If you want to reduce the score of fg_thr, which file needs to be modified?

**Comment by GuoxingYan** [comment]
@makefile @Peng-wei-Yu 
When I was training, batchsize was equal to 1. There was at least one sample in my own training pictures, but Why is total positive equal to 0 in many iterations during the training process?and my rpn loss is 0.Have you encountered such a problem?
![default](https://user-images.githubusercontent.com/32195410/41136889-f07956a8-6b0a-11e8-9735-06081c73a0a1.png)

**Comment by GuoxingYan** [comment]
@makefile Did you try to change the short_size and long_size in train.prototxt？when i only changed the short_size or long_size ,There will be an error。

**Comment by makefile** [comment]
@GuoxingYan I did not try to change that, since there use Deconvolution layer to upsample, the size maybe need to be multiplier of 32, 64 or larger.

**Comment by GuoxingYan** [comment]
@makefile thank you very much！！

**Comment by GuoxingYan** [comment]
@makefile Will you have the following problems when training fpn?
![default](https://user-images.githubusercontent.com/32195410/41692344-59f30026-7532-11e8-97cd-2b017255cbd1.png)

**Comment by makefile** [comment]
@GuoxingYan I didn't met. the integer seems to be abnormal big.

**Comment by licy5152** [comment]
@Peng-wei-Yu @zhaoweicai  my own data size is 960*1280,I try to use the ResNet-50-model-merge.caffemodel, but I also get this problem.
<img width="604" alt="wx20180624-154016 2x" src="https://user-images.githubusercontent.com/30167614/41816992-eefa0720-77c4-11e8-88a7-e0e0ed66973b.png">

**Comment by GuoxingYan** [comment]
@makefile @zhaoweicai @Peng-wei-Yu When I was training, I found that the short_size in detection_data_param in trian.prototxt is 800, which is exactly equal to img_width and img_height in proposal_target_param. So the question arises. When I change the short_size to 320, does the img_width and img_height need to be changed to 320?

**Comment by makefile** [comment]
@GuoxingYan I think it needs to be.

**Comment by licy5152** [comment]
@makefile  I use to train my owe dataset,how can I get the output for every picture?

**Comment by makefile** [comment]
@licy5152 I wrote a python script [CascadeRCNN-demo.py](https://gist.github.com/makefile/6731ca0e311b6401681c15635bb97330) imitate the matlab code, you can modify it to use.

**Comment by GuoxingYan** [comment]
@makefile 你的demo.py 显示无效链接诶。

**Comment by makefile** [comment]
@GuoxingYan 你的网络问题吧

**Comment by PacteraKun** [comment]
@makefile @zhaoweicai 
When I was training my own dataset, the following issue happened. However, I have already check that there is no box has xmin = 1664 and xmax = 636 in the window_file.txt. And I also have not found bbox_util.cpp file under the workspace directory.  Could you guys help me to solve this issue? Thanks a lot. 
![image](https://user-images.githubusercontent.com/40010065/42417380-6c991a9a-823d-11e8-9ccb-77f87325fd9c.png)

**Comment by makefile** [comment]
@PacteraKun The situation you encountered is unusual, check carefully.

**Comment by PacteraKun** [comment]
@makefile 
Have you use cascade-rcnn  to train your own dataset successfully?

**Comment by makefile** [comment]
@PacteraKun I once trained several model, but failed to visualize the demo result. Later I transplant it to my own familiar framework to use.

**Comment by lzh19961031** [comment]
@makefile 请问下，你test那个python文件中的labelmap_file是什么呢？

**Comment by GuoxingYan** [comment]
@emmra https://blog.csdn.net/e01528/article/details/80913443 希望能帮到你，可以的话，帮忙点个赞。

**Comment by GuoxingYan** [comment]
@emmra 保存不了caffemodel的问题我没有遇到

**Comment by GuoxingYan** [comment]
@lzh19961031 那个检测的python你能打开吗？
我这边试了好几次没有打开，方便的话可以给我 发一下吗？
921905071@qq.com

**Comment by huinsysu** [comment]
@makefile @GuoxingYan @licy5152 请问，在train.prototext文件中的long_size和short_size的作用是什么呢？我得数据集中有的图片长宽分别为6000和4000,我需要在设置这两个参数为6000和4000吗？谢谢！

**Comment by huinsysu** [comment]
@makefile @zhaoweicai Hi,I try to lower the fg_th in  BoxGroupOutput layer, but I still get the problem of `keep_num > 0(0 vs. 0).` Could I just set the fg_th to 0 for all fg_th in  BoxGroupOutput layer? Thanks for help me!

**Comment by makefile** [comment]
@huinsysu the size is about input resize. 6000x4000 maybe too large to fit into 1 gpu.

**Comment by huinsysu** [comment]
@licy5152 Hi, when I trained the model with my own dataset, I met the same error as you met. Would you please tell me how you solve such problem? Thanks！

**Comment by hu5tao** [comment]
@makefile At last,are you satisfied with you results about your datasets? I am preparing for train my dataset in my datasets.

**Comment by makefile** [comment]
@hu5tao not bad.

**Comment by qianfangjj** [comment]
> @licy5152 I wrote a python script [CascadeRCNN-demo.py](https://gist.github.com/makefile/6731ca0e311b6401681c15635bb97330) imitate the matlab code, you can modify it to use.

@makefile  你好，我试了好几次都不能打开 CascadeRCNN-demo.py的链接，请问你是否方便发给我一份？422246019@qq.com 谢谢了！

**Comment by foralliance** [comment]
@lininglouis 
0.99??
I think its just drop cls_score is lower than 0.01

**Comment by leizhu1989** [comment]
when I train my own data ,it has a error,but I don't know why,could you give me some ideas? Thanks a lot

I0604 13:28:15.270220 87804 detection_data_layer.cpp:142] num: 0 /home/zhulei/data/VOCdevkit/VOC2007/JPEGImages/IMG_0_112.jpg 3 1080 1920 windows to process: 36, RONI windows: 0
F0604 13:28:15.274016 87804 detection_data_layer.cpp:123] Check failed: label > 0 (0 vs. 0) 
*** Check failure stack trace: ***
    @     0x7fb962af05cd  google::LogMessage::Fail()
    @     0x7fb962af2433  google::LogMessage::SendToLog()
    @     0x7fb962af015b  google::LogMessage::Flush()
    @     0x7fb962af2e1e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7fb963271781  caffe::DetectionDataLayer<>::DataLayerSetUp()
    @     0x7fb9631c27d5  caffe::BasePrefetchingDataLayer<>::LayerSetUp()
    @     0x7fb96338b6a2  caffe::Net<>::Init()
    @     0x7fb96338dd0e  caffe::Net<>::Net()
    @     0x7fb963312515  caffe::Solver<>::InitTrainNet()
    @     0x7fb963312aa4  caffe::Solver<>::Init()
    @     0x7fb963312d8f  caffe::Solver<>::Solver()
    @     0x7fb963335701  caffe::Creator_SGDSolver<>()
    @           0x40d912  train()
    @           0x408795  main
    @     0x7fb961335830  __libc_start_main
    @           0x4090a9  _start
    @              (nil)  (unknown)
